In [ ]:
import os

In [ ]:
mainpath = '/p/home/jusers/cherepashkin1/jureca/cherepashkin1/598csv'

In [ ]:
#if not(os.path.isfile('tmean.csv') and os.path.isfile(mainpath+'/sh_paramters.csv')):
def rewalk(mainpath, lastfilename):
    cip = []
    for root, directories, filenames in os.walk(mainpath): 
        for filename in filenames:
            if filename[-len(lastfilename):] == lastfilename:
                cip.append(os.path.join(root,filename))
    return(cip)

In [ ]:
cip = []
lastfilename = 'Y_N'
for root, directories, filenames in os.walk(mainpath): 
    for filename in filenames:
        if lastfilename in filename:
            os.remove(os.path.join(root,filename))

In [ ]:
cip = []
lastfilename = 'bX'
for root, directories, filenames in os.walk(mainpath): 
    for filename in filenames:
        if lastfilename in filename:
            os.remove(os.path.join(root,filename))

In [ ]:
import torch
import numpy as np
import os

In [ ]:
y_n = np.genfromtxt(os.path.join('/p/home/jusers/cherepashkin1/jureca/cherepashkin1/',\
                                              str(598)+'csv','Y_N_'+str(1000)+'.csv').replace('\\','/'),delimiter=',')

In [ ]:
y_n.shape

In [ ]:
y_n = np.expand_dims(y_n, axis = 0)

In [ ]:
y_n = np.repeat(y_n, 40, axis = 0)

In [ ]:
y_n.shape

In [ ]:
loss_between = 'f+f_n'

In [ ]:
f = []
l = []
if loss_between == 'f' or loss_between == 'f+f_n':
    f = 1
elif loss_between == 'f_n':
    l = 0
if loss_between == 'f':
    l = f
elif loss_between == 'f+f_n':
    l = 2*f

In [ ]:
print(f,l)

In [ ]:
y_n = torch.zeros([1,500, 441])
y_n = y_n.repeat(40,1,1)
y_n.shape

In [ ]:

outputs = torch.zeros([40, 441])
f = torch.zeros([40,500])
for i in range(outputs.shape[0]):
    f[i] = torch.matmul(y_n[i],outputs[i])

In [ ]:
y_n = torch.zeros([40, 500, 441])
outputs = torch.zeros([40, 441])
#f = torch.zeros([40,500])
f = torch.matmul(y_n,outputs)

In [ ]:
f.shape

# making frame csv with pathes to the F_N files

In [ ]:
import pandas as pd

In [ ]:
mainpath = '/p/home/jusers/cherepashkin1/jureca/cherepashkin1/'
specie = 598
nsp = 500
for dtype in ['train', 'val']:
    csvpath = os.path.join(mainpath,str(specie)+'csv',dtype)
    cip = []
    for root, directories, filenames in os.walk(csvpath): 
        for filename in filenames:
            search_str = 'F_N_'+str(nsp)+'.csv'
            if filename[-len(search_str):] == search_str:
                twofolds = os.path.join(root, filename).split('/')[-2:]
                cip.append(os.path.join(twofolds[0],twofolds[1]))
    #                 cip.append(os.path.join(root, filename).split('/')[-2:-1])
    lframe = pd.DataFrame()
    lframe.insert(0, 'file_name', \
                           [cip[i].split('_F_N')[0] for i in range(len(cip))])
    lframe.to_csv(os.path.join(csvpath, 'f_n_folders_pathes_'+str(nsp)+dtype+'.csv'), index=False)

In [ ]:
root_dir = csvpath
print(root_dir)
folder_names_frame = pd.read_csv(os.path.join(root_dir,'f_n_folders_pathes_'+str(nsp)+root_dir.split('/')[-1]+'.csv'))

In [ ]:
folder_names_frame

In [ ]:
specie = '598,619'
sl = specie.split(',')

In [ ]:
specie = '598,619'
nsp = 500
mainpath = '/p/home/jusers/cherepashkin1/jureca/cherepashkin1/phenoseed_csv'
cip = []
for word in specie.split(','):
    csvpath = os.path.join(mainpath,word+'csv')
    for root, directories, filenames in os.walk(csvpath): 
        for filename in filenames:
            search_str = 'F_N_'+str(nsp)+'.csv'
            if filename[-len(search_str):] == search_str:
                tfolds = os.path.join(root, filename).split('/')[-3:]
                cip.append(os.path.join(tfolds[0],tfolds[1],tfolds[2]))
lframe = pd.DataFrame()
lframe.insert(0, 'file_name', \
                       [cip[i].split('_F_N')[0] for i in range(len(cip))])
lframe = lframe.sample(frac=1)
#lframe.to_csv(pts, index=False)

In [ ]:
df = [lframe[:100], lframe[100:]]

In [ ]:
df[1]

In [ ]:
cip[-1]

In [37]:
for root, directories, filenames in os.walk('/p/home/jusers/cherepashkin1/jureca/cherepashkin1/phenoseed_csv/619csv'): 
    for filename in filenames:
#         search_str = 'F_N_'+str(nsp)+'.csv'
        if 'Y_N_500.csv' in filename:
            os.remove(os.path.join(root, filename))
#         if filename[-len(search_str):] == search_str:
#             tfolds = os.path.join(root, filename).split('/')[-3:]

In [41]:
import torch
import numpy as np

In [42]:
    tmean = torch.Tensor(np.genfromtxt('/p/home/jusers/cherepashkin1/jureca/circles/finetune_test/tmean.csv', delimiter = ',')).cuda()

In [46]:
outputs = torch.zeros(40,441).cuda()

In [47]:
ampl = 441
outputs = torch.multiply(outputs,tmean[3,:ampl]-tmean[2,:ampl])+tmean[2,:ampl]

# Checking that datafolders have _Surface.ply

In [3]:
import os

In [1]:
cip = []
mainpath = '\\ibg2fs6.ibg.kfa-juelich.de\phenoseed$\phenoseed_data'

In [6]:
os.path.isfile(mainpath)

False

In [5]:
mainpath = '/p/home/jusers/cherepashkin1/jureca/cherepashkin1/phenoseed'

In [7]:
os.chdir(mainpath)

In [11]:
cip = []
cip2 = []
for root, directories, filenames in os.walk(mainpath): 
    for filename in filenames:
#         search_str = 'F_N_'+str(nsp)+'.csv'
        if 'Surface.ply' in filename:
#             print(root.split('/')[8])
            cip2.append(os.path.join(root, filename))
#             cip.append(root.split('/')[3].split('\\')[0])
            cip.append(root.split('/')[8])
#         if filename[-len(search_str):] == search_str:
#             tfolds = os.path.join(root, filename).split('/')[-3:]

In [15]:
len(cip)

28

In [16]:
cip

['628',
 '646',
 '642',
 '649',
 '650',
 '630',
 '658',
 '617',
 '660',
 '652',
 '603',
 '653',
 '615',
 '593',
 '644',
 '647',
 '655',
 '592',
 '1445909',
 '632',
 '598',
 '594',
 '654',
 '648',
 '600',
 '656',
 '619',
 '645']

In [14]:
cip = list(set(cip))

In [18]:
cip.sort()

In [19]:
cip

['1445909',
 '592',
 '593',
 '594',
 '598',
 '600',
 '603',
 '615',
 '617',
 '619',
 '628',
 '630',
 '632',
 '642',
 '644',
 '645',
 '646',
 '647',
 '648',
 '649',
 '650',
 '652',
 '653',
 '654',
 '655',
 '656',
 '658',
 '660']

In [13]:
len(cip2)

14235

In [23]:
'/'+j+'/' in i

False

In [24]:
dc = {}
for i in cip2:
    for j in cip:
        if '/'+j+'/' in i:
            dc[j]= dc.get(j,0)+1

In [28]:
sort_orders = sorted(dc.items(), key=lambda x: x[1], reverse=True)

for i in sort_orders:
	print(i[0], i[1])

598 5283
603 4261
654 501
645 386
648 384
646 384
644 356
649 288
600 288
630 240
592 216
593 216
594 216
615 216
656 191
617 162
619 108
647 96
628 96
655 95
652 50
650 48
658 48
642 42
653 32
1445909 15
632 9
660 8


In [27]:
sorted(dc)

['1445909',
 '592',
 '593',
 '594',
 '598',
 '600',
 '603',
 '615',
 '617',
 '619',
 '628',
 '630',
 '632',
 '642',
 '644',
 '645',
 '646',
 '647',
 '648',
 '649',
 '650',
 '652',
 '653',
 '654',
 '655',
 '656',
 '658',
 '660']

In [20]:
dc = {}
for i in cip2:
    for j in cip:
        if '/'+j+'/' in cip2:
            dc[j]+=1

In [21]:
dc

{}

In [10]:
list(set(cip))

['598']

# coping only ply, prmat and rot img from 603

In [ ]:
 Namespace(ampl=441, bn=10000, bs=35, center=False, chidden_dim='chidden_dim = np.hstack((96,128,np.repeat(256, 3)))', classicnorm=False, cmscrop=550, criterion='L1', csvname='598csv11', downsample=1, epoch=200, expdescr='truly normalized dataset', expnum='e054', feature_extract=False, gradient_predivide_factor=1.0, haf=True, hidden_dim='hidden_dim = np.hstack((np.repeat(32, 1),1500))', inputt='img', kernel_sizes='kernel_sizes = np.hstack((7,np.repeat(3, 5)))', lb='pc', lr=0.01, machine='workstation', maintain=False, maintain_line=True, man_dist=False, merging_order='', minmax=False, minmax3dimage=False, minmax_f=False, model_name='', netname='cnet', ngpu=4, noise_input=False, noise_output=False, normalize=False, num_input_images=3, num_sam_points=500, num_workers=0, outputt='single_f_n', parallel='torch', pc_scale=5.0, pin_memory=False, rand_angle=False, rescale=550, rot_dirs=False, save_output=False, single_folder=False, specie='598', use_adasum=False, use_cuda=False, use_existing_csv=True, use_pretrained=False, use_sep_csv=True, view_sep=True, wandb='', weight_decay=0, zero_angle=False)


In [ ]:
import wandb

# Checking that datafolders have .ply

In [1]:
import os

In [1]:
cip = []
mainpath = '\\ibg2fs6.ibg.kfa-juelich.de\phenoseed$\phenoseed_data'

In [6]:
os.path.isfile(mainpath)

False

In [2]:
mainpath = '/p/home/jusers/cherepashkin1/jureca/cherepashkin1/phenoseed'

In [6]:
'jureca' not in mainpath

False

In [3]:
os.chdir(mainpath)

In [29]:
rotar = ['rotation_'+str(10*i).zfill(3)+'.tif' for i in range(36)]

In [35]:
all(['rotation_'+str(10*i).zfill(3)+'.tif' in filename for i in range(36)])

False

In [33]:
all([True, True, True])

True

In [30]:
rotar

['rotation_000.tif',
 'rotation_010.tif',
 'rotation_020.tif',
 'rotation_030.tif',
 'rotation_040.tif',
 'rotation_050.tif',
 'rotation_060.tif',
 'rotation_070.tif',
 'rotation_080.tif',
 'rotation_090.tif',
 'rotation_100.tif',
 'rotation_110.tif',
 'rotation_120.tif',
 'rotation_130.tif',
 'rotation_140.tif',
 'rotation_150.tif',
 'rotation_160.tif',
 'rotation_170.tif',
 'rotation_180.tif',
 'rotation_190.tif',
 'rotation_200.tif',
 'rotation_210.tif',
 'rotation_220.tif',
 'rotation_230.tif',
 'rotation_240.tif',
 'rotation_250.tif',
 'rotation_260.tif',
 'rotation_270.tif',
 'rotation_280.tif',
 'rotation_290.tif',
 'rotation_300.tif',
 'rotation_310.tif',
 'rotation_320.tif',
 'rotation_330.tif',
 'rotation_340.tif',
 'rotation_350.tif']

In [31]:
cip = []
cip2 = []
for root, directories, filenames in os.walk(mainpath): 
    for filename in filenames:
#         search_str = 'F_N_'+str(nsp)+'.csv'
        if '.ply' in filename and 'Surface' not in filename and
            all(['rotation_'+str(10*i).zfill(3)+'.tif' in filename for i in range(36)]):
#             print(root.split('/')[8])
            cip2.append(os.path.join(root, filename))
#             cip.append(root.split('/')[3].split('\\')[0])
            cip.append(root.split('/')[8])
#         if filename[-len(search_str):] == search_str:
#             tfolds = os.path.join(root, filename).split('/')[-3:]

KeyboardInterrupt: 

In [8]:
len(cip)

31939

In [16]:
cip

['628',
 '646',
 '642',
 '649',
 '650',
 '630',
 '658',
 '617',
 '660',
 '652',
 '603',
 '653',
 '615',
 '593',
 '644',
 '647',
 '655',
 '592',
 '1445909',
 '632',
 '598',
 '594',
 '654',
 '648',
 '600',
 '656',
 '619',
 '645']

In [9]:
cip = list(set(cip))

In [10]:
cip.sort()

In [11]:
cip

['1',
 '1445909',
 '316',
 '319',
 '327',
 '328',
 '329',
 '330',
 '333',
 '335',
 '336',
 '337',
 '339',
 '341',
 '347',
 '350',
 '351',
 '352',
 '363',
 '364',
 '365',
 '367',
 '371',
 '374',
 '413',
 '414',
 '415',
 '416',
 '417',
 '418',
 '421',
 '422',
 '424',
 '425',
 '426',
 '430',
 '437',
 '439',
 '440',
 '470',
 '471',
 '476',
 '478',
 '479',
 '480',
 '481',
 '495',
 '502',
 '503',
 '504',
 '507',
 '508',
 '509',
 '511',
 '513',
 '515',
 '516',
 '517',
 '518',
 '522',
 '523',
 '524',
 '529',
 '530',
 '531',
 '532',
 '540',
 '541',
 '592',
 '593',
 '594',
 '598',
 '600',
 '603',
 '615',
 '617',
 '619',
 '628',
 '630',
 '632',
 '642',
 '644',
 '645',
 '646',
 '647',
 '648',
 '649',
 '650',
 '652',
 '653',
 '654',
 '655',
 '656',
 '658',
 '660']

In [12]:
len(cip2)

31939

In [23]:
'/'+j+'/' in i

False

In [13]:
dc = {}
for i in cip2:
    for j in cip:
        if '/'+j+'/' in i:
            dc[j]= dc.get(j,0)+1

In [14]:
sort_orders = sorted(dc.items(), key=lambda x: x[1], reverse=True)

for i in sort_orders:
	print(i[0], i[1])

598 5283
603 4261
600 1758
532 1392
530 1296
531 1248
439 1212
481 1035
440 999
511 958
656 763
417 578
509 577
508 573
654 501
655 467
645 386
648 384
541 384
646 384
644 356
480 353
649 288
422 288
479 274
365 241
630 240
339 240
592 216
593 216
594 216
615 216
513 200
515 200
516 196
421 194
367 193
364 192
517 192
518 192
416 191
418 191
617 162
333 146
371 120
328 120
347 120
619 108
415 100
414 98
504 96
647 96
524 96
628 96
529 95
351 79
503 79
352 76
471 51
652 50
540 49
650 48
658 48
478 45
642 42
507 41
653 32
329 26
327 24
425 24
426 24
363 23
523 22
413 20
437 19
522 19
1445909 16
502 15
319 12
341 11
632 9
660 8
337 8
330 5
470 5
374 5
350 5
495 4
335 4
476 4
430 2
316 2
1 2
424 2
336 2


In [16]:
import numpy as np

In [18]:
arr = np.array(list(dc.items()))

In [19]:
arr.shape

(95, 2)

In [23]:
arr = arr.astype(int)

In [24]:
np.sum(arr,0)

array([1491646,   31939])

In [ ]:
with h5py.File('groups.hdf5', 'w') as f:
    g = f.create_group('Base_Group')
    gg = g.create_group('Sub_Group')
    d = g.create_dataset('default', data=arr)
    dd = gg.create_dataset('default', data=arr)

In [36]:
import h5py
import helpers
import numpy as np
from pathlib import Path
import torch
from torch.utils import data

class HDF5Dataset(data.Dataset):
    """Represents an abstract HDF5 dataset.
    
    Input params:
        file_path: Path to the folder containing the dataset (one or multiple HDF5 files).
        recursive: If True, searches for h5 files in subdirectories.
        load_data: If True, loads all the data immediately into RAM. Use this if
            the dataset is fits into memory. Otherwise, leave this at false and 
            the data will load lazily.
        data_cache_size: Number of HDF5 files that can be cached in the cache (default=3).
        transform: PyTorch transform to apply to every data instance (default=None).
    """
    def __init__(self, file_path, recursive, load_data, data_cache_size=3, transform=None):
        super().__init__()
        self.data_info = []
        self.data_cache = {}
        self.data_cache_size = data_cache_size
        self.transform = transform

        # Search for all h5 files
        p = Path(file_path)
        assert(p.is_dir())
        if recursive:
            files = sorted(p.glob('**/*.h5'))
        else:
            files = sorted(p.glob('*.h5'))
        if len(files) < 1:
            raise RuntimeError('No hdf5 datasets found')

        for h5dataset_fp in files:
            self._add_data_infos(str(h5dataset_fp.resolve()), load_data)
            
    def __getitem__(self, index):
        # get data
        x = self.get_data("data", index)
        if self.transform:
            x = self.transform(x)
        else:
            x = torch.from_numpy(x)

        # get label
        y = self.get_data("label", index)
        y = torch.from_numpy(y)
        return (x, y)

    def __len__(self):
        return len(self.get_data_infos('data'))
    
    def _add_data_infos(self, file_path, load_data):
        with h5py.File(file_path) as h5_file:
            # Walk through all groups, extracting datasets
            for gname, group in h5_file.items():
                for dname, ds in group.items():
                    # if data is not loaded its cache index is -1
                    idx = -1
                    if load_data:
                        # add data to the data cache
                        idx = self._add_to_cache(ds.value, file_path)
                    
                    # type is derived from the name of the dataset; we expect the dataset
                    # name to have a name such as 'data' or 'label' to identify its type
                    # we also store the shape of the data in case we need it
                    self.data_info.append({'file_path': file_path, 'type': dname, 'shape': ds.value.shape, 'cache_idx': idx})

    def _load_data(self, file_path):
        """Load data to the cache given the file
        path and update the cache index in the
        data_info structure.
        """
        with h5py.File(file_path) as h5_file:
            for gname, group in h5_file.items():
                for dname, ds in group.items():
                    # add data to the data cache and retrieve
                    # the cache index
                    idx = self._add_to_cache(ds.value, file_path)

                    # find the beginning index of the hdf5 file we are looking for
                    file_idx = next(i for i,v in enumerate(self.data_info) if v['file_path'] == file_path)

                    # the data info should have the same index since we loaded it in the same way
                    self.data_info[file_idx + idx]['cache_idx'] = idx

        # remove an element from data cache if size was exceeded
        if len(self.data_cache) > self.data_cache_size:
            # remove one item from the cache at random
            removal_keys = list(self.data_cache)
            removal_keys.remove(file_path)
            self.data_cache.pop(removal_keys[0])
            # remove invalid cache_idx
            self.data_info = [{'file_path': di['file_path'], 'type': di['type'], 'shape': di['shape'], 'cache_idx': -1} if di['file_path'] == removal_keys[0] else di for di in self.data_info]

    def _add_to_cache(self, data, file_path):
        """Adds data to the cache and returns its index. There is one cache
        list for every file_path, containing all datasets in that file.
        """
        if file_path not in self.data_cache:
            self.data_cache[file_path] = [data]
        else:
            self.data_cache[file_path].append(data)
        return len(self.data_cache[file_path]) - 1

    def get_data_infos(self, type):
        """Get data infos belonging to a certain type of data.
        """
        data_info_type = [di for di in self.data_info if di['type'] == type]
        return data_info_type

    def get_data(self, type, i):
        """Call this function anytime you want to access a chunk of data from the
            dataset. This will make sure that the data is loaded in case it is
            not part of the data cache.
        """
        fp = self.get_data_infos(type)[i]['file_path']
        if fp not in self.data_cache:
            self._load_data(fp)
        
        # get new cache_idx assigned by _load_data_info
        cache_idx = self.get_data_infos(type)[i]['cache_idx']
        return self.data_cache[fp][cache_idx]

ModuleNotFoundError: No module named 'helpers'

In [1]:
import time

In [7]:
print(time.ctime())

Fri Nov  5 21:08:09 2021
